content/articles/email-parser-python/index.md

```diff
@@ -0,0 +1,172 @@
+      #parse using the defaul line break(\n)
+        return email.parser.BytesParser(policy=email.policy.default).parse(f)
+#load
+normal_emails = [load_emails(is_spam=False, filename=name) for name in normal_filenames]
+spam_emails = [load_emails(is_spam=True, filename=name) for name in spam_filenames]
+```
+
+We create a directory each for the spam and normal emails and sort them.
+
+In the `load_emails()`, we open the appropriate directories as readable and in binary format, then we parse them using the `BytesParser` class.
+
+The `BytesParser` class contains an argument in the constructor called `policy`. It has the policy as default. Using `default` lets us parse the email using the `\n` line breaks.
+
+We then call the `load_emails()` method to load the emails.
+
+We can look at a sample of a normal email in the next cell.
+
+```python
+print(normal_emails[42].get_content().strip())
+```
+
+#### Viewing the structures of the emails
+
+In this part, we will look at how to get the structure of the emails and further look at the common types of structures for spam and normal emails.
+
+Paste this in the next cell.
+
+```python
+'''Getting the most common email structures'''
+def get_structures(email):
+  #if its plain text return text/plain
+    if isinstance(email, str):
+        return email
+    email_payload = email.get_email_payload()
+    #if the payload is a list then its probably a multipart
+    #return a multipart thereafter
+    if isinstance(email_payload, list):
+        return "multipart({})".format(", ".join([
+            get_structures(sub_email)
+            for sub_email in email_payload
+        ]))
+    else:
+        return email.get_content_type()
+        
+#function for counting the types
+def type_counter(emails):
+    our_count = Counter()
+    for email in emails:
+        structure = get_structures(email)
+        our_count[structure] += 1
+    return our_count
+```
+
+In this code, we have two methods, `get_structures()` and `type_counter()`.
+
+In the `get_structures()` function, we check the structure. If it is a normal text email, we return `text/plain,` but if it's a multipart type of email, we return multipart, all the parts it contains. Note also that recursion is used if there are many email structures in the sub-emails of that email.
+For any other email structure we display it by return the emails content type(`get_content_type()`).
+
+For the `type_counter()` method, we check how many email structure types are gotten in the spam and normal emails, e.g. `'text/plain', 2409`. So we first initiate a counter then count them for every similar structure.
+
+We can display one for normal emails using this code:
+
+```python
+print(structures_counter(normal_emails))
+```
+
+We check the most common using the `most_common()` method.
+
+```python
+print(structures_counter(spam_emails).most_common())
+```
+ We should see the output as shown below:
```

Comment:
Instead of just pasting the output, let us know why we are seeing what we are seeing. Do this regularly for all outputs that you want to discuss.